src/pghstore/_speedups.c

```diff
@@ -0,0 +1,373 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <stdio.h>
+#include <string.h>
+
+struct module_state {
+    PyObject *error;
+};
+
+#if PY_MAJOR_VERSION >= 3
+#define GETSTATE(m) ((struct module_state*)PyModule_GetState(m))
+#endif
+
+#define CHAR_CHECK(s, pos, c)                        \
+  if (s[pos] != char) {                              \
+    PyErr_Format(PyExc_ValueError,                   \
+                 "Encountered %c at %i, was expecting %c.",   \
+                 s[pos], pos, c);                             \
+    return NULL;                                              \
+  }
+
+#define isEven(a) (((a) & 1) == 0)
+
+#define ConcatOrGotoCleanup(result, string) { PyBytes_Concat(&result, string); if (!result) { goto _speedup_dumps_cleanup_and_exit; } }
+
+/*
+ * Just like strchr, but find first -unescaped- occurrence of c in s.
+ */
+char *
+strchr_unescaped(char *s, char c) 
+{
+  char *p = strchr(s, c), *q;
+  while (p != NULL) { /* loop through all the c's */
+    q = p; /* scan backwards through preceding escapes */
+    while (q > s && *(q-1) == '\\')
+      q--;
+    if (isEven(p-q)) /* even number of esc's => c is good */
+        return p;
+    p = strchr(p+1, c); /* else odd escapes => c is escaped, keep going */
+  }
+  return NULL;
+}
+
+/* Written to match pghstore._native.unescape but slightly smarter. */
+char *
+unescape(char *start_of_string, char *end_of_string)
+{
+  ssize_t copy_index = 0;
+  ssize_t index;
+  ssize_t string_length = (end_of_string - start_of_string) + 1;
+  /* NOTE(sigmavirus24): In the event that there are \s in the string, we're
+   * over-allocating here but that is okay since we're going to free shortly
+   * after use.
+   */
+  char *unescaped_string = malloc(string_length);
+  memset(unescaped_string, 0, string_length);
```

Comment:
https://stackoverflow.com/questions/1538420/difference-between-malloc-and-calloc